[libpas] Fix potential null dereferences in PGM crash reporter

zakariaridouh opened the pull request:
#### b6f1cd836ae3ee46aee384ab80c2c85c94724246
<pre>
[libpas] Fix potential null dereferences in PGM crash reporter
<a href="https://bugs.webkit.org/show_bug.cgi?id=298407">https://bugs.webkit.org/show_bug.cgi?id=298407</a>
<a href="https://rdar.apple.com/159871696">rdar://159871696</a>

Reviewed by Yusuke Suzuki.

We should do some additional checks on pointers before dereferencing them.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_report_crash_extract_pgm_failure):

Canonical link: <a href="https://commits.webkit.org/300253@main">https://commits.webkit.org/300253@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/308b4e331154d47cd0abfc84f00fc3487de8b430

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119977 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30321 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126307 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72050 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ffb87258-d277-4da5-af0e-659b05ab3da8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40366 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91112 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60424 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4909182f-7fa0-4290-92c7-bdde3d1e1e43) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32240 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107582 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71667 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6da92884-2d48-49b5-9953-9c7f238338b0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69942 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/112102 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101715 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129227 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118493 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46897 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35562 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99739 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103770 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99584 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25629 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45029 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23046 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43518 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46759 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/147192 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46225 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37821 "Found 1 new JSC binary failure: testapi, Found 18636 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/StackArgsWithFormals.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply3.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49574 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47911 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -87,7 +90,7 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
     kern_return_t kr = reader(task, pas_dead_root, sizeof(pas_root), (void**)&read_pas_dead_root);
-    if (kr != KERN_SUCCESS)
+    if (kr != KERN_SUCCESS || !read_pas_dead_root)
         return KERN_FAILURE;
```
I would like to keep it separate since they are different reasons.

```
if (kr != KERN_SUCCESS)
    return KERN_FAILURE;
if (!read_pas_dead_root)
    return KERN_FAILURE;
```

zakariaridouh replied:
That's fair, updating PR.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -77,6 +77,9 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
 {
+    if (!report || !crm_reader)
+        return KERN_INVALID_ARGUMENT;
+
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -99,27 +102,31 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
-    if (kr != KERN_SUCCESS)
+    if (kr != KERN_SUCCESS || !temp_pgm_has_been_used_ptr)
         return KERN_FAILURE;
+    report->pgm_has_been_used = *temp_pgm_has_been_used_ptr;
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -99,27 +102,31 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
     kr = reader(task, (vm_address_t)read_pas_dead_root->pas_pgm_hash_map_instance, sizeof(pas_ptr_hash_map), (void**)&hash_map);
-    if (kr != KERN_SUCCESS)
+    if (kr != KERN_SUCCESS || !hash_map)
         return KERN_FAILURE;
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -99,27 +102,31 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
         kr = reader(task, (vm_address_t)(hash_map->table + i), sizeof(pas_ptr_hash_map_entry), (void**)&hash_map_entry);
-        if (kr != KERN_SUCCESS)
+        if (kr != KERN_SUCCESS || !hash_map_entry)
             return KERN_FAILURE;
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -99,27 +102,31 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
         kr = reader(task, (vm_address_t)hash_map_entry->value, sizeof(pas_pgm_storage), (void**)&pgm_metadata);
-        if (kr != KERN_SUCCESS)
+        if (kr != KERN_SUCCESS || !pgm_metadata)
             return KERN_FAILURE;
```
Ditto.

zakariaridouh committed:
[libpas] Fix potential null dereferences in PGM crash reporter
https://bugs.webkit.org/show_bug.cgi?id=298407
rdar://159871696

Reviewed by Yusuke Suzuki.

We should do some additional checks on pointers before dereferencing them.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_report_crash_extract_pgm_failure):

Canonical link: https://commits.webkit.org/300253@main
